mesonbuild/interpreter/interpreter.py

```diff
@@ -1534,6 +1534,10 @@ def add_find_program_override(self, name: str, exe: T.Union[build.Executable, Ex
             raise InterpreterException(f'Tried to override finding of executable "{name}" which has already been found.')
         if name in self.build.find_overrides:
             raise InterpreterException(f'Tried to override executable "{name}" which has already been overridden.')
+        if isinstance(exe, build.Executable) and exe.for_machine == MachineChoice.HOST and \
+           not self.environment.can_run_host_binaries():
+            mlog.warning('Override program with an executable that cannot be run.',
+                         location=self.current_node)
```

Comment:
It is very much actionable by the project, and what's more, it's ***required*** for either:
- the project to workaround it in cross compile cases
- meson to refuse to override it when it cannot work

You make a fair point for option 2, and in that case I agree it should be informational. Let's do that.

Reply:
Right now, the only way for a project to "workaround it in cross compile cases" is to either:

1. Set `native: true` on the executable then make that work by duplicating the entire dependency tree of that executable with `native: true`, or
2. Wrap it in `if meson.can_run_host_binaries()` which is an obvious fix that is both trivial to know when to do and has no edge-cases where it's wrong, so meson should just do it automatically. This is good behaviour, and it should be logged so that users have transparency into what is happening.

Note that we actually have no way to have wraps provide programs conditionally. So if we add a warning, we're just forcing needless churn in projects, especially since we ideally want every project that installs things into bindir to provide it. Maybe that is even something we should do fully automatically (with just a wrap `[provide]` entry, no `meson.override_find_program()`).

About (1), I checked in the gstreamer monorepo (which has a number of wraps) and in wrapdb, and I found zero cases where just setting `native: true` on an `executable()` would suffice. All cases had internal dependencies, and often some external ones. It's just too much unnecessary work to duplicate the dependency tree and maintain correctness (it increases the testing area, making it harder for people get it right).

Honestly, I think the whole `native: true` functionality has always been incomplete. We need a way for meson to _automatically_ duplicate the entire dependency tree for native compilation. This would work for internal deps, but we would need a fully-declarative external dependency syntax, or more kwargs for `dependency()` calls to work in more cases. We've discussed that in the past, but that's a digression here.

Reply:
> ... Unless there are cases where the user would rather things error out with an un-runnable program instead of getting buggy build artifacts with a build machine program that doesn't produce the same results as a cross compiled one?

I can't think of a case like this where you wouldn't just use the built executable directly. Why look at `PATH` at all?

Reply:
> Maybe that is even something we should do fully automatically (with just a wrap `[provide]` entry, no `meson.override_find_program()`).

Absolutely not. The majority of programs aren't even useful as buildsystem dependencies. For example, all GUI programs without exception. Also some projects are weird and can only be built with an older version of that very project -- compilers like rust for example. We can say that's a bad idea, but making it an irreconcilable dependency loop is a bit much. What about cases where something like a `sh` implementation relies on the build machine having the posix-mandated `sh` installed, but wants to build a new one, or a competing implementation, and has to generate scripted sources?

> I can't think of a case like this where you wouldn't just use the built executable directly. Why look at `PATH` at all?

I thought it was fairly obvious that this hypothetical involves a subproject, and using the subproject built program in the main project.

Reply:
> Absolutely not. The majority of programs aren't even useful as buildsystem dependencies.

If someone writes a wrap file with a `[provide]` section with a list of binaries they want that subproject to provide, they're explicitly opting in to that program being overriden. The rest of your paragraph doesn't apply then. We can match the name in the `[provide]` section to the name of the `executable()` directly. `meson.override_find_program()` is redundant. It only exists because it predates the concept of a `[provide]` section.

> I thought it was fairly obvious that this hypothetical involves a subproject, and using the subproject built program in the main project as a fallback or a user-added wrap not shipped with the upstream tarball.

I was talking about `get_variable()`, and I'm disappointed that you resorted to condescension rather than considering that I may have a point that you didn't consider. Anyway, my point is you either:

1. Really just want to use what you are building directly and you don't want anything else, in which case why go through a global namespace (`PATH`) to use `[provide]` + `meson.override_find_program()` + `find_program()` and then on top of that apply redundant versioning and sourcing checks to your executable, or
2. You have a wayland-scanner like situation, where the producer and consumer are separate entities, and the producer really wants a mechanism to ensure that the dependency object and the program go together. Providing them both separately is impossible to get right, since we don't provide a way for the user to communicate their intent to Meson. The wayland developers have argued the same, and I agree with them.

(2) is the same problem as (1), really, and completely separate from this situation.